File: 2_medium/03517_smallest_palindromic_rearrangement_i/solution.cpp
class Solution {
public:
    string smallestPalindrome(string s) {
        const int SIZE = 26;
        int freq[SIZE]{};

        for (auto ch: s) {
            freq[ch - 'a'] += 1;
        }

        int l{0}, r{(int)s.size() - 1};

        for (int i{0}; i < SIZE; ++i) {

            if ((freq[i] & 1) == 1) {
                int mid = s.size() / 2;
                s[mid] = i + 'a';
            }

            for (int j{0}; j < freq[i]/2; ++j) {
                s[l] = i + 'a';
                s[r] = i + 'a';
                l += 1;
                r -= 1;
            }
        }

        return s;
    }
};

File: 2_medium/03517_smallest_palindromic_rearrangement_i/solution.rs
impl Solution {
    pub fn smallest_palindrome(mut s: String) -> String {       
        let mut s = s.into_bytes();
        let mut freq = [0i32;26];

        s.iter().for_each(|&ch| freq[(ch - b'a') as usize] += 1);

        let mut l = 0;
        let mut r = s.len() - 1;

        for i in 0..freq.len() {

            if (freq[i] & 1) == 1 {
                let mid = s.len() / 2;
                s[mid] = i as u8 + b'a';
            }

            for j in 0..freq[i]/2 {
                s[l] = i as u8 + b'a';
                s[r] = i as u8 + b'a';
                l += 1;
                r -= 1;
            }
        }

        String::from_utf8(s).unwrap()
    }
}
